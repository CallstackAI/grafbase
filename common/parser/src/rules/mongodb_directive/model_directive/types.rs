pub(crate) mod create;
pub(crate) mod delete;
pub(crate) mod filter;
pub(crate) mod generic;
pub(crate) mod oneof;
