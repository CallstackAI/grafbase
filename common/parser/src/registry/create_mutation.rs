use crate::rules::relations::generate_metarelation;
use crate::rules::visitor::VisitorContext;
use crate::utils::{is_modelized_node, to_base_type_str, to_defined_input_type, to_input_type, to_lower_camelcase};
use case::CaseExt;
use dynaql::indexmap::IndexMap;
use dynaql::registry::relations::MetaRelation;
use dynaql::registry::{
    resolvers::dynamo_mutation::DynamoMutationResolver, resolvers::dynamo_querying::DynamoResolver,
    resolvers::Resolver, resolvers::ResolverType, variables::VariableResolveDefinition, MetaField, MetaInputValue,
    MetaType,
};
use dynaql::Operations;
use dynaql::Positioned;
use dynaql_parser::types::{FieldDefinition, ObjectType, TypeDefinition, TypeKind};

/// Create an input type for a Node's Relation Create.
///
/// ```graphql
/// input PostPublishedAuthorCreateInput {
///   create: ...AuthorWithoutPublishedRelation
///   link: ID
/// }
/// ```
fn create_input_relation<'a>(
    ctx: &mut VisitorContext<'a>,
    ty_from: &TypeDefinition,
    ty_to: &TypeDefinition,
    relation: &MetaRelation,
    field: &Positioned<FieldDefinition>,
) -> String {
    let ty_from_name = ty_from.name.node.to_camel();
    let ty_to_name = ty_to.name.node.to_camel();

    let prefix = format!(
        "{}{}{}",
        ty_from_name.to_camel(),
        relation.name.to_camel(),
        ty_to_name.to_camel(),
    );

    let input_name = format!("{prefix}CreateInput");

    if ctx.types.get(&input_name).is_some() {
        return input_name;
    }

    match &ty_to.kind {
        TypeKind::Object(object) => {
            let mut input_fields = IndexMap::new();
            for field in &object.fields {
                let name = &field.node.name.node;
                // If it's a modelized node, we want to generate
                let actual_field_type = is_modelized_node(&ctx.types, &field.node.ty.node);

                if actual_field_type.is_some() {
                    let relation_name = generate_metarelation(ty_to, &field.node).name;
                    if relation_name == relation.name {
                        // If we are in the same relation we try to reverse
                        continue;
                    }

                    let field_base_ty = to_base_type_str(&field.node.ty.node.base);
                    let input_name = format!(
                        "{}{}{}CreateInput",
                        &ty_to.name.node.to_camel(),
                        relation_name.clone().to_camel(),
                        field_base_ty.to_camel()
                    );

                    input_fields.insert(
                        name.to_string(),
                        MetaInputValue {
                            name: name.to_string(),
                            description: field.node.description.clone().map(|x| x.node),
                            ty: to_defined_input_type(field.node.ty.clone().node, input_name).to_string(),
                            visible: None,
                            default_value: None,
                            is_secret: false,
                        },
                    );
                    continue;
                }

                // TODO: Abstract this behind an `ID` utility;
                if name.ne("id") {
                    input_fields.insert(
                        name.clone().to_string(),
                        MetaInputValue {
                            name: name.to_string(),
                            description: field.node.description.clone().map(|x| x.node),
                            ty: to_input_type(&ctx.types, field.node.ty.clone().node).to_string(),
                            visible: None,
                            default_value: None,
                            is_secret: false,
                        },
                    );
                    continue;
                }
            }

            ctx.registry.get_mut().create_type(
                &mut |_| MetaType::InputObject {
                    name: input_name.clone(),
                    description: Some(format!("Input to create a new {prefix}")),
                    oneof: false,
                    input_fields: input_fields.clone(),
                    visible: None,
                    rust_typename: input_name.clone(),
                },
                &input_name,
                &input_name,
            );
        }
        _ => ctx.report_error(
            vec![field.pos],
            format!(
                "You should have an `Object` type here for field: `{}` in `{}`.",
                field.node.name.node, &ty_from.name.node
            ),
        ),
    }

    let input_name_link = format!("{prefix}CreateRelationInput");

    ctx.registry.get_mut().create_type(
        &mut |_| MetaType::InputObject {
            name: input_name_link.clone(),
            description: Some(format!("Input to create a new {prefix} relation")),
            oneof: true,
            input_fields: {
                let mut input_fields = IndexMap::new();

                input_fields.insert(
                    "create".to_string(),
                    MetaInputValue {
                        name: "create".to_string(),
                        description: None,
                        ty: input_name.clone(),
                        visible: None,
                        default_value: None,
                        is_secret: false,
                    },
                );

                input_fields.insert(
                    "link".to_string(),
                    MetaInputValue {
                        name: "link".to_string(),
                        description: None,
                        ty: "ID".to_string(),
                        visible: None,
                        default_value: None,
                        is_secret: false,
                    },
                );

                input_fields
            },
            visible: None,
            rust_typename: input_name_link.clone(),
        },
        &input_name_link,
        &input_name_link,
    );

    input_name_link
}

/// We do create the `input` type of every possibility for a Type.
///
/// For each `@modelized` directive, we want to create an `input` available
/// for each type.
///
/// # Example
///
/// ```graphql
/// type Post @modelized {
///   id: ID!
///   author: Author @relation(name: "published")
///   comments: [Comment] @relation(name: "comments")
///   ...
/// }
/// ```
///
/// Would create
///
/// ```graphql
/// input PostPublishedAuthorCreateRelationInput {...}
/// input PostCommentsCommentCreateRelationInput {...}
/// ```
pub fn create_input_without_relation<'a>(ctx: &mut VisitorContext<'a>, ty: &TypeDefinition, object: &ObjectType) {
    let type_name = ty.name.node.to_camel();
    let create_input_name = format!("{}CreateInput", type_name);
    let mut input_fields = IndexMap::new();

    if ctx.types.get(&create_input_name).is_some() {
        return;
    }

    for field in &object.fields {
        let name = &field.node.name.node;
        // If it's a modelized node, we want to generate
        let types = ctx.types.clone(); // TODO: We should change a little the way it works, this clone can be avoided, not really expensive but should still be reworked.

        let actual_field_type = is_modelized_node(&types, &field.node.ty.node);
        if let Some(ty_to) = actual_field_type {
            // Should trigger the creation of the sub input
            let relation = generate_metarelation(ty, &field.node);
            let input_name = create_input_relation(ctx, ty, &ty_to.node, &relation, field);

            input_fields.insert(
                name.to_string(),
                MetaInputValue {
                    name: name.to_string(),
                    description: field.node.description.clone().map(|x| x.node),
                    ty: to_defined_input_type(field.node.ty.clone().node, input_name).to_string(),
                    visible: None,
                    default_value: None,
                    is_secret: false,
                },
            );
            continue;
        }

        // TODO: Abstract this behind an `ID` utility;
        if name.ne("id") {
            input_fields.insert(
                name.clone().to_string(),
                MetaInputValue {
                    name: name.to_string(),
                    description: field.node.description.clone().map(|x| x.node),
                    ty: to_input_type(&ctx.types, field.node.ty.clone().node).to_string(),
                    visible: None,
                    default_value: None,
                    is_secret: false,
                },
            );
            continue;
        }
    }

    ctx.registry.get_mut().create_type(
        &mut |_| MetaType::InputObject {
            name: create_input_name.clone(),
            description: Some(format!("Input to create a new {}", &type_name)),
            oneof: false,
            input_fields: input_fields.clone(),
            visible: None,
            rust_typename: type_name.clone(),
        },
        &create_input_name,
        &create_input_name,
    );
}

/// The idea there is to generate the create mutation of an Entity depending on
/// the fields of the Entity. If it's linked to another Node based on a relation
/// we'll generate an Input based on a `link` or a `create`.
///
/// ```graphql
/// type Post {
///   id: ID!
///   content: String
///   author: Author @relation(name: "published")
///   comments: [Comment] @relation(name: "comments")
///   ...
/// }
///
/// type Author {
///   id: ID!
///   firstName: String
///   lastName: String
///   postPublished: [Post] @relation(name: "published")
///   comments: [Comment] @relation(name: "commented")
/// }
///
/// type Comment {
///   id: ID!
///   content: String
///   post: Post @relation(name: "published")
///   author: Author @relation(name: "commented")
/// }
/// ```
///
/// Would create something like:
///
/// ```graphql
/// input PostPublishedCreateRelationInput {
///   """
///   As the relation `published` got an edge from `Author` to `Post`, we do not
///   need to have this relation available on the creation.
///   """
///   create: PostPublishedAuthorCreateRelationInput
///   link: ID
/// }
///
/// """
/// This is the Author Input type without the `published` relation for the `Post`
/// """
/// type PostPublishedAuthorCreateRelationInput {
///   firstName: String
///   lastName: String
///   comments: [AuthorCommentedCommentCreateRelationInput]
/// }
///
/// """
/// This is the Comment Input type without the `commented` relation for the `Author`
/// """
/// type AuthorCommentedCommentCreateRelationInput {
///   content: String
///   post: CommentPublishedPostCreateRelationInput
/// }
///
/// type PostCommentCreateRelationInput {
/// }
///
/// """
/// Post create Input type
/// """
/// input PostCreateInput {
///   content: String
///   author: PostPublishedCreateRelationInput
///   comments: [PostCommentCreateRelationInput]
/// }
///
/// type Mutation {
///   postCreate(input: PostCreateInput): PostPayload
/// }
/// ```
///
pub fn add_create_mutation<'a>(
    ctx: &mut VisitorContext<'a>,
    ty: &TypeDefinition,
    object: &ObjectType,
    type_name: &str,
) {
    create_input_without_relation(ctx, ty, object);
    let type_name = type_name.to_string();
    let create_input_name = format!("{}CreateInput", type_name.to_camel());

    let create_payload_name = format!("{}CreatePayload", type_name.to_camel());
    // CreatePayload
    ctx.registry.get_mut().create_type(
        &mut |_| MetaType::Object {
            name: create_payload_name.clone(),
            description: None,
            fields: {
                let mut fields = IndexMap::new();
                let name = to_lower_camelcase(&type_name);
                fields.insert(
                    name.clone(),
                    MetaField {
                        name,
                        description: None,
                        args: Default::default(),
                        ty: type_name.to_camel(),
                        deprecation: Default::default(),
                        cache_control: Default::default(),
                        external: false,
                        requires: None,
                        provides: None,
                        visible: None,
                        compute_complexity: None,
                        edges: Vec::new(),
                        relation: None,
                        resolve: Some(Resolver {
                            id: Some(format!("{}_resolver", type_name.to_lowercase())),
                            // Single entity
                            r#type: ResolverType::DynamoResolver(DynamoResolver::QueryPKSK {
                                pk: VariableResolveDefinition::LocalData("id".to_string()),
                                sk: VariableResolveDefinition::LocalData("id".to_string()),
                            }),
                        }),
                        transforms: None,
                        required_operation: Some(Operations::CREATE),
                    },
                );
                fields
            },
            cache_control: dynaql::CacheControl {
                public: true,
                max_age: 0usize,
            },
            extends: false,
            keys: None,
            visible: None,
            is_subscription: false,
            is_node: false,
            rust_typename: create_payload_name.clone(),
            constraints: vec![],
        },
        &create_payload_name,
        &create_payload_name,
    );

    // createQuery
    ctx.mutations.push(MetaField {
        name: format!("{}Create", to_lower_camelcase(&type_name)),
        description: Some(format!("Create a {}", type_name)),
        args: {
            let mut args = IndexMap::new();
            args.insert(
                "input".to_owned(),
                MetaInputValue {
                    name: "input".to_owned(),
                    description: None,
                    ty: format!("{}!", &create_input_name),
                    default_value: None,
                    visible: None,
                    is_secret: false,
                },
            );
            args
        },
        ty: create_payload_name,
        deprecation: dynaql::registry::Deprecation::NoDeprecated,
        cache_control: dynaql::CacheControl {
            public: true,
            max_age: 0usize,
        },
        external: false,
        provides: None,
        requires: None,
        visible: None,
        edges: Vec::new(),
        relation: None,
        compute_complexity: None,
        resolve: Some(Resolver {
            id: Some(format!("{}_create_resolver", type_name.to_lowercase())),
            r#type: ResolverType::DynamoMutationResolver(DynamoMutationResolver::CreateNode {
                input: VariableResolveDefinition::InputTypeName("input".to_owned()),
                ty: type_name,
            }),
        }),
        transforms: None,
        required_operation: Some(Operations::CREATE),
    });
}

/*
 * TODO: Fix this
#[cfg(test)]
mod tests {
    use dynaql::{Name, Pos, Positioned, Schema};
    use dynaql_parser::types::{FieldDefinition, ObjectType, ServiceDocument, Type, TypeDefinition};
    use insta::assert_snapshot;

    use crate::rules::visitor::VisitorContext;

    use super::add_create_mutation;

    #[test(skip)]
    fn ensure_create_mutation_types() {
        let doc = ServiceDocument { definitions: vec![] };
        let mut ctx = VisitorContext::new(&doc);
        let field_definition = FieldDefinition {
            ty: Positioned::new(Type::new("Author").unwrap(), Pos::default()),
            description: None,
            name: Positioned::new(Name::new("Author"), Pos::default()),
            arguments: Vec::new(),
            directives: Vec::new(),
        };

        let fake_object_ty = ObjectType {
            fields: vec![Positioned {
                pos: Pos { line: 3, column: 4 },
                node: field_definition,
            }],
            implements: Vec::new(),
        };

        let fake_object_type_def = TypeDefinition {
            kind: dynaql_parser::types::TypeKind::Object(fake_object_ty.clone()),
            description: None,
            name: Positioned {
                pos: Pos { line: 1, column: 2 },
                node: Name::new("Author"),
            },
            directives: Vec::new(),
            extend: false,
        };

        add_create_mutation(&mut ctx, &fake_object_type_def, &fake_object_ty, "Author");
        let sdl = Schema::new(ctx.registry.take()).sdl();
        assert_snapshot!(sdl);
    }
}
*/
