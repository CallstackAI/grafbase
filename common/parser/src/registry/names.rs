use crate::{registry::ParentRelation, utils::to_lower_camelcase};
use case::CaseExt;
use dynaql_parser::types::TypeDefinition;

use super::NumericFieldKind;

pub const PAGINATION_INPUT_ARG_FIRST: &str = "first";
pub const PAGINATION_INPUT_ARG_LAST: &str = "last";
pub const PAGINATION_INPUT_ARG_BEFORE: &str = "before";
pub const PAGINATION_INPUT_ARG_AFTER: &str = "after";

pub const INPUT_ARG_BY: &str = "by";
pub const INPUT_ARG_INPUT: &str = "input";

pub const INPUT_FIELD_RELATION_CREATE: &str = "create";
pub const INPUT_FIELD_RELATION_LINK: &str = "link";
pub const INPUT_FIELD_RELATION_UNLINK: &str = "unlink";
pub const INPUT_FIELD_NUM_OP_SET: &str = "set";
pub const INPUT_FIELD_NUM_OP_INCREMENT: &str = "increment";
pub const INPUT_FIELD_NUM_OP_DECREMENT: &str = "decrement";

pub struct MetaNames;

/// Defines the names used by the different generated types.
/// It looks a bit silly to centralize names, but they're part of the public API and they SHOULD be
/// consistent.
impl MetaNames {
    // FIXME: Several places used to_camel() but not everywhere... Do we want to enforce it?
    pub fn model(model_type_definition: &TypeDefinition) -> String {
        model_type_definition.name.node.to_camel()
    }

    pub fn mutation_create(model_type_definition: &TypeDefinition) -> String {
        to_lower_camelcase(format!("{}Create", Self::model(model_type_definition)))
    }

    pub fn mutation_update(model_type_definition: &TypeDefinition) -> String {
        to_lower_camelcase(format!("{}Update", Self::model(model_type_definition)))
    }

    pub fn mutation_create_payload_type(model_type_definition: &TypeDefinition) -> String {
        format!("{}CreatePayload", Self::model(model_type_definition))
    }

    pub fn mutation_update_payload_type(model_type_definition: &TypeDefinition) -> String {
        format!("{}UpdatePayload", Self::model(model_type_definition))
    }

    /// Defines
    /// - without parent, the create mutation input type name.
    /// - with parent, the nested input type name to create said type when creating the parent.
    pub fn create_input_type(
        model_type_definition: &TypeDefinition,
        maybe_parent_relation: Option<&ParentRelation<'_>>,
    ) -> String {
        format!(
            "{}CreateInput",
            maybe_parent_relation
                .map(|parent_relation| Self::relation_prefix(parent_relation, model_type_definition))
                .unwrap_or_else(|| Self::model(model_type_definition))
        )
    }

    pub fn update_input_type(model_type_definition: &TypeDefinition) -> String {
        format!("{}UpdateInput", Self::model(model_type_definition))
    }

    /// For a given relation, one can either link to an existing object or create a new one.
    pub fn create_relation_input_type(
        parent_relation: &ParentRelation<'_>,
        field_model_type_definition: &TypeDefinition,
    ) -> String {
        format!(
            "{}CreateRelationInput",
            Self::relation_prefix(parent_relation, field_model_type_definition)
        )
    }

    /// For a given relation, one can either change the (un)link to an existing object or create a new one
    pub fn update_relation_input_type(
        parent_relation: &ParentRelation<'_>,
        field_model_type_definition: &TypeDefinition,
    ) -> String {
        format!(
            "{}UpdateRelationInput",
            Self::relation_prefix(parent_relation, field_model_type_definition)
        )
    }

    pub fn numerical_operation(kind: &NumericFieldKind) -> String {
        format!("{}OperationsInput", kind.as_str())
    }

    /// Prefix used for any input/output type created for a relation.
    fn relation_prefix(parent_relation: &ParentRelation<'_>, field_model_type_definition: &TypeDefinition) -> String {
        format!(
            "{}{}{}",
            &Self::model(parent_relation.model_type_definition),
            &parent_relation.meta.name.to_camel(),
            Self::model(field_model_type_definition)
        )
    }
}
