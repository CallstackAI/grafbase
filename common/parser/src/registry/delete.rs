use dynaql::registry::transformers::Transformer;
use dynaql::registry::{
    resolvers::dynamo_mutation::DynamoMutationResolver, resolvers::Resolver, resolvers::ResolverType,
    variables::VariableResolveDefinition, MetaField, MetaInputValue,
};
use dynaql::registry::{InputObjectType, NamedType, ObjectType};

use dynaql::{AuthConfig, CacheControl};
use dynaql_parser::types::{BaseType, TypeDefinition};
use grafbase::auth::Operations;

use crate::registry::names::{MetaNames, INPUT_ARG_BY, INPUT_ARG_INPUT};
use crate::rules::visitor::VisitorContext;
use crate::type_names::TypeNameExt;

pub fn add_mutation_delete<'a>(
    ctx: &mut VisitorContext<'a>,
    model_type_definition: &'a TypeDefinition,
    model_auth: Option<&AuthConfig>,
    cache_control: CacheControl,
) {
    let type_name = MetaNames::model(model_type_definition);
    let payload = register_payload(ctx, model_type_definition, model_auth, cache_control.clone());

    // deleteMutation
    ctx.mutations.push(MetaField {
        name: MetaNames::mutation_delete(model_type_definition),
        description: Some(format!("Delete a {type_name} by ID or unique field")),
        args: [MetaInputValue::new(
            INPUT_ARG_BY,
            format!("{}!", MetaNames::by_input(model_type_definition)),
        )]
        .into_iter()
        .map(|input| (input.name.clone(), input))
        .collect(),
        ty: payload.as_nullable().into(),
        resolve: Some(Resolver {
            id: Some(format!("{}_delete_resolver", type_name.to_lowercase())),
            r#type: ResolverType::DynamoMutationResolver(DynamoMutationResolver::DeleteNode {
                ty: type_name.clone().into(),
                by: VariableResolveDefinition::InputTypeName(INPUT_ARG_BY.to_owned()),
            }),
        }),

        cache_control: cache_control.clone(),
        required_operation: Some(Operations::DELETE),
        auth: model_auth.cloned(),
        ..Default::default()
    });

    let many_input_type = register_many_input(ctx, model_type_definition);
    let delete_many_payload = register_many_payload(ctx, model_type_definition, model_auth, cache_control.clone());
    ctx.mutations.push(MetaField {
        name: MetaNames::mutation_delete_many(model_type_definition),
        description: Some(format!("Delete multiple {type_name}")),
        args: [MetaInputValue::new(INPUT_ARG_INPUT, format!("[{many_input_type}!]!"))]
            .into_iter()
            .map(|input| (input.name.clone(), input))
            .collect(),
        ty: delete_many_payload.as_nullable().into(),
        resolve: Some(Resolver {
            id: None,
            r#type: ResolverType::DynamoMutationResolver(DynamoMutationResolver::DeleteNodes {
                input: VariableResolveDefinition::InputTypeName(INPUT_ARG_INPUT.to_owned()),
                ty: type_name.into(),
            }),
        }),
        cache_control,
        required_operation: Some(Operations::DELETE),
        auth: model_auth.cloned(),
        ..Default::default()
    });
}

fn register_many_input(ctx: &mut VisitorContext<'_>, model_type_definition: &TypeDefinition) -> BaseType {
    let input_type_name = MetaNames::delete_many_input(model_type_definition);

    ctx.registry.borrow_mut().create_type(
        |_| {
            InputObjectType::new(
                input_type_name.clone(),
                [MetaInputValue::new(
                    INPUT_ARG_BY,
                    format!("{}!", MetaNames::by_input(model_type_definition)),
                )],
            )
            .into()
        },
        &input_type_name,
        &input_type_name,
    );

    BaseType::named(&input_type_name)
}

fn register_payload(
    ctx: &mut VisitorContext<'_>,
    model_type_definition: &TypeDefinition,
    model_auth: Option<&AuthConfig>,
    cache_control: CacheControl,
) -> NamedType<'static> {
    let payload_type_name = MetaNames::delete_payload_type(model_type_definition);
    // DeletePayload
    ctx.registry.get_mut().create_type(
        |_| {
            ObjectType::new(
                payload_type_name.clone(),
                [MetaField {
                    name: "deletedId".to_string(),
                    ty: NamedType::from("ID").as_non_null().into(),
                    transformer: Some(Transformer::JSONSelect {
                        property: "id".to_string(),
                    }),
                    required_operation: Some(Operations::DELETE),
                    auth: model_auth.cloned(),
                    cache_control: cache_control.clone(),
                    ..Default::default()
                }],
            )
            .into()
        },
        &payload_type_name,
        &payload_type_name,
    );

    payload_type_name.into()
}

fn register_many_payload(
    ctx: &mut VisitorContext<'_>,
    model_type_definition: &TypeDefinition,
    model_auth: Option<&AuthConfig>,
    cache_control: CacheControl,
) -> NamedType<'static> {
    let payload_type_name = MetaNames::delete_many_payload_type(model_type_definition);
    // DeletePayload
    ctx.registry.get_mut().create_type(
        |_| {
            ObjectType::new(
                payload_type_name.clone(),
                [MetaField {
                    name: "deletedIds".to_string(),
                    ty: NamedType::from("ID").as_non_null().list().non_null().into(),
                    transformer: Some(Transformer::JSONSelect {
                        property: "ids".to_string(),
                    }),
                    required_operation: Some(Operations::DELETE),
                    auth: model_auth.cloned(),
                    cache_control: cache_control.clone(),
                    ..Default::default()
                }],
            )
            .into()
        },
        &payload_type_name,
        &payload_type_name,
    );

    payload_type_name.into()
}
