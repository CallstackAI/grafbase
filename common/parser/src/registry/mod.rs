//! TODO:
//!
//! -> Split each of the creation and add tests with SDL
//!
use case::CaseExt;

use dynaql::indexmap::IndexMap;
use dynaql::registry::enums::DynaqlEnum;
use dynaql::registry::relations::MetaRelation;
use dynaql::registry::transformers::Transformer;
use dynaql::registry::{
    resolvers::dynamo_mutation::DynamoMutationResolver, resolvers::Resolver, resolvers::ResolverType,
    variables::VariableResolveDefinition, MetaField, MetaInputValue, MetaType,
};
use dynaql::registry::{MetaEnumValue, Registry};
use dynaql::validation::dynamic_validators::DynValidator;
use dynaql::AuthConfig;
use dynaql_parser::types::{BaseType, FieldDefinition, ObjectType, TypeDefinition};
use grafbase::auth::Operations;
use std::fmt::Display;

use crate::registry::names::MetaNames;
use crate::rules::length_directive::{LENGTH_DIRECTIVE, MAX_ARGUMENT, MIN_ARGUMENT};
use crate::rules::visitor::VisitorContext;
use crate::utils::{to_input_type, to_lower_camelcase};

mod mutations;
pub mod names;
mod pagination;
mod relations;
mod search;
pub use mutations::{add_mutation_create, add_mutation_update, NumericFieldKind};
pub use pagination::{add_query_paginated_collection, generate_pagination_args};
pub use search::add_query_search;

fn register_dynaql_enum<T: DynaqlEnum>(registry: &mut Registry) -> BaseType {
    let type_name = T::ty().to_string();
    registry.create_type(
        |_| MetaType::Enum {
            name: type_name.clone(),
            description: None,
            enum_values: IndexMap::from_iter(
                T::values()
                    .into_iter()
                    .map(|value| {
                        (
                            value.clone(),
                            MetaEnumValue {
                                name: value,
                                description: None,
                                deprecation: dynaql::registry::Deprecation::NoDeprecated,
                                visible: None,
                                value: None,
                            },
                        )
                    })
                    .collect::<Vec<_>>(),
            ),
            visible: None,
            rust_typename: type_name.clone(),
        },
        &type_name,
        &type_name,
    );
    BaseType::named(&type_name)
}

/// Create an input type for a non_primitive Type.
pub fn add_input_type_non_primitive(ctx: &mut VisitorContext<'_>, object: &ObjectType, type_name: &str) -> String {
    let type_name = type_name.to_string();
    let input_type = format!("{}Input", type_name.to_camel());

    // Input
    ctx.registry.get_mut().create_type(
        |_| MetaType::InputObject {
            name: input_type.clone(),
            description: Some(format!("{type_name} input type.")),
            oneof: false,
            input_fields: {
                let mut input_fields = IndexMap::new();
                for field in &object.fields {
                    let name = &field.node.name.node;

                    input_fields.insert(
                        name.clone().to_string(),
                        MetaInputValue {
                            description: field.node.description.clone().map(|x| x.node),
                            ..MetaInputValue::new(
                                name.to_string(),
                                to_input_type(&ctx.types, field.node.ty.clone().node),
                            )
                        },
                    );
                }
                input_fields
            },
            visible: None,
            rust_typename: input_type.clone(),
        },
        &input_type,
        &input_type,
    );

    input_type
}

/// Add the remove mutation for a given Object
pub fn add_remove_mutation(ctx: &mut VisitorContext<'_>, type_name: &str, auth: Option<&AuthConfig>) {
    let type_name = type_name.to_string();
    let delete_payload_name = format!("{}DeletePayload", type_name.to_camel());

    // DeletePayload
    ctx.registry.get_mut().create_type(
        |_| MetaType::Object {
            name: delete_payload_name.clone(),
            description: None,
            fields: {
                let mut fields = IndexMap::new();
                let name = "deletedId".to_string();
                fields.insert(
                    name.clone(),
                    MetaField {
                        name,
                        description: None,
                        args: Default::default(),
                        // TODO: Should be infered from the entity depending on the directives
                        ty: "ID!".to_string(),
                        deprecation: Default::default(),
                        cache_control: Default::default(),
                        external: false,
                        requires: None,
                        provides: None,
                        visible: None,
                        compute_complexity: None,
                        edges: Vec::new(),
                        relation: None,
                        resolve: None,
                        transformer: Some(Transformer::JSONSelect {
                            property: "id".to_string(),
                        }),
                        plan: None,
                        required_operation: Some(Operations::DELETE),
                        auth: auth.cloned(),
                    },
                );
                fields
            },
            cache_control: Default::default(),
            extends: false,
            keys: None,
            is_node: false,
            visible: None,
            is_subscription: false,
            rust_typename: delete_payload_name.clone(),
            constraints: vec![],
        },
        &delete_payload_name,
        &delete_payload_name,
    );

    // deleteMutation
    ctx.mutations.push(MetaField {
        name: format!("{}Delete", to_lower_camelcase(&type_name)),
        description: Some(format!("Delete a {type_name} by ID or unique field")),
        args: {
            let mut args = IndexMap::new();
            args.insert(
                "by".to_owned(),
                MetaInputValue::new("by", format!("{type_name}ByInput!")),
            );
            args
        },
        ty: delete_payload_name,
        deprecation: dynaql::registry::Deprecation::NoDeprecated,
        cache_control: Default::default(),
        external: false,
        provides: None,
        requires: None,
        visible: None,
        edges: Vec::new(),
        relation: None,
        compute_complexity: None,
        resolve: Some(Resolver {
            id: Some(format!("{}_delete_resolver", type_name.to_lowercase())),
            r#type: ResolverType::DynamoMutationResolver(DynamoMutationResolver::DeleteNode {
                ty: type_name,
                by: VariableResolveDefinition::InputTypeName("by".to_owned()),
            }),
        }),
        plan: None,
        transformer: None,
        required_operation: Some(Operations::DELETE),
        auth: auth.cloned(),
    });
}

fn get_length_validator(field: &FieldDefinition) -> Option<DynValidator> {
    use tuple::Map;
    field
        .directives
        .iter()
        .find(|directive| directive.node.name.node == LENGTH_DIRECTIVE)
        .map(|directive| {
            let (min_value, max_value) = (MIN_ARGUMENT, MAX_ARGUMENT).map(|argument_name| {
                directive.node.get_argument(argument_name).and_then(|argument| {
                    if let dynaql_value::ConstValue::Number(ref min) = argument.node {
                        min.as_u64().and_then(|min| min.try_into().ok())
                    } else {
                        None
                    }
                })
            });
            DynValidator::length(min_value, max_value)
        })
}

/// Used to keep track of the parent relation when created nested input types
/// TODO: Merge it with MetaRelation?
pub struct ParentRelation<'a> {
    /// TypeDefinition of @model type
    model_type_definition: &'a TypeDefinition,
    meta: &'a MetaRelation,
}

impl<'a> Display for ParentRelation<'a> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "{} relation of {}",
            self.meta.name,
            MetaNames::model(self.model_type_definition)
        )
    }
}
