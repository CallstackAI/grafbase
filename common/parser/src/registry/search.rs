use dynaql::indexmap::IndexMap;
use dynaql::registry::resolvers::dynamo_querying::DynamoResolver;
use dynaql::registry::resolvers::query::{QueryResolver, SearchField, SearchSchema, MATCHING_RECORDS_ID_KEY};
use dynaql::registry::{
    resolvers::Resolver, resolvers::ResolverType, variables::VariableResolveDefinition, MetaField, MetaInputValue,
};
use dynaql::{AuthConfig, Operations};
use dynaql_parser::types::{ConstDirective, FieldDefinition, TypeDefinition};

use crate::registry::names::MetaNames;
use crate::rules::visitor::VisitorContext;

const SEARCH_INPUT_ARG_QUERY: &str = "query";
const SEARCH_INPUT_ARG_LIMIT: &str = "limit";

pub fn add_query_search(
    ctx: &mut VisitorContext<'_>,
    model_type_definition: &TypeDefinition,
    model_auth: Option<&AuthConfig>,
    search_fields: Vec<(&FieldDefinition, &ConstDirective)>,
) {
    assert!(!search_fields.is_empty());
    let type_name = MetaNames::model(model_type_definition);
    let field_name = MetaNames::search(model_type_definition);

    let search_schema = SearchSchema {
        fields: search_fields
            .into_iter()
            .map(|(field, _directive)| SearchField {
                name: field.name.node.to_string(),
                r#type: field.ty.node.to_string(),
            })
            .collect(),
    };

    ctx.queries.push(MetaField {
        name: field_name,
        description: Some(format!("Search `{type_name}`")),
        args: IndexMap::from([
            (
                SEARCH_INPUT_ARG_QUERY.to_string(),
                MetaInputValue {
                    name: SEARCH_INPUT_ARG_QUERY.to_string(),
                    description: Some("Raw search query".to_string()),
                    ty: "String!".to_string(),
                    default_value: None,
                    visible: None,
                    validators: None,
                    is_secret: false,
                },
            ),
            (
                SEARCH_INPUT_ARG_LIMIT.to_string(),
                MetaInputValue {
                    name: SEARCH_INPUT_ARG_LIMIT.to_string(),
                    description: Some("Maximum number of documents to retrieve (defaults to 10)".to_string()),
                    ty: "Int!".to_string(),
                    default_value: Some(10.into()),
                    visible: None,
                    validators: None,
                    is_secret: false,
                },
            ),
        ]),
        ty: format!("[{type_name}!]!"),
        deprecation: dynaql::registry::Deprecation::NoDeprecated,
        cache_control: dynaql::CacheControl {
            public: true,
            max_age: 0usize,
        },
        external: false,
        provides: None,
        requires: None,
        visible: None,
        compute_complexity: None,
        edges: vec![],
        relation: None,
        plan: None,
        resolve: Some(
            Resolver {
                id: None,
                r#type: ResolverType::Query(QueryResolver::Search {
                    query: VariableResolveDefinition::InputTypeName(SEARCH_INPUT_ARG_QUERY.to_string()),
                    limit: VariableResolveDefinition::InputTypeName(SEARCH_INPUT_ARG_LIMIT.to_string()),
                    r#type: type_name,
                    schema: search_schema,
                }),
            }
            .and_then(ResolverType::DynamoResolver(DynamoResolver::QueryIds {
                ids: VariableResolveDefinition::LocalData(MATCHING_RECORDS_ID_KEY.to_string()),
            })),
        ),
        transformer: None,
        required_operation: Some(Operations::LIST),
        auth: model_auth.cloned(),
    });
}
