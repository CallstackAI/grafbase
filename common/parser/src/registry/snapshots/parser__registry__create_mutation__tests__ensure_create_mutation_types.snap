---
source: common/parser/src/registry/create_mutation.rs
assertion_line: 518
expression: sdl
---
type AuthorCreatePayload {
	author: Author
}
"""
Input to create a new Author
"""
input AuthorCreationInput {
	Author: Error
}
schema {
	query: Query
}

